src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -427,6 +439,13 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             );
                         }
 
+                        if is_fn_trait && is_target_feature_fn {
+                            err.note(&format!(
+                                "`{}` has `#[target_feature]` and is unsafe to call",
+                                trait_ref.skip_binder().self_ty(),
```

Comment:
...and then here we can use the `def-id` to give an error like `foo` has `#[target_feature]`, instead of `fn() {foo}`, which is really unclear.

Also, I think "unsafe to call" is perhaps a bit misleading. Maybe just "`foo` has `#[target_feature]` annotations and therefore does not implement the `Fn` traits" or something like that?

Reply:
Looking at the error text again, I'm not sure it's even necessary to repeat the name of the function, since it's in the help text as well (albeit with the somewhat confusing `fn() {foo}`).  In my latest commit I've changed it to simply "`#[target_feature]` functions do not implement the `Fn` traits".  Do you think it's necessary to specify the function name in the error note?

Reply:
No, I don't.